### PR TITLE
feat (reader): Add option to add custom bottom padding to footnote popup

### PR DESCRIPTION
**Pull request by cookiebit**
Quite a lot of phones now have rounded screen corners, meaning that content all the way in the corner risks ending up outside of the physical screen. This happens with footnote popups, where a character or two in the two bottom corners ends up not being shown on screen, or only partially.

This problem can be solved by adding an adjustable bottom margin to the popup widget, so that all the content is pushed a given amount away from the screen edge.

I have made a proof of concept, as there are some things I am not very familiar with in the codebase, so I'd appreciate if someone has any pointers to what could be improved. For instance, I'm not exactly sure the units used are the best ones for the purpose, and if that logic is implemented correctly, although it seems to work as a concept at least. Also, the help texts etc. could probably be improved?

I have no experience building koreader from source though, so I have only tried the frontend changes on the source code in the zip file in release 2025.10.

The added settings:
New menu entries

Widget to set the custom value:
Widget

Before (no extra padding):
Before

After:
After

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/14866)

### CODE REVIEW

**Comment by poire-z** [comment]
Makes sense.

> I have no experience building koreader from source though, so I have only tried the frontend changes on the source code in the zip file in release 2025.10.

That's totally fine, for little changes, that's how everybody started I guess. (Once you make larger things and you're tired having to push your change to the device, launch koreader, crash, look at the crash.log, fix, and do that again, you'll love the emulator).

> For instance, I'm not exactly sure the units used are the best ones for the purpose, and if that logic is implemented correctly, although it seems to work as a concept at least. Also, the help texts etc. could probably be improved?

Most of our units are "unit less", some number without unit, that we scaleBySize() to make it pixels. What you did seems fine.

I would probably just get rid of the toggle, and just have, under "Footnote popup size": "Foonote bottom padding" which launches the SpinWidget with a default of 0 and an added button to reset to 0.
It's not really something that once configured (if you need 60px on your phone) you will want to toggle.

File: frontend/ui/widget/footnotewidget.lua [hunk 1]
```diff
@@ -308,6 +308,11 @@ function FootnoteWidget:init()
     local padding_top = Size.padding.large
     local padding_bottom = Size.padding.large
     local htmlwidget_height = self.height - padding_top - padding_bottom
+    if G_reader_settings:isTrue("footnote_popup_add_bottom_padding") then
+        local padding_bottom_extra =  Screen:scaleBySize(G_reader_settings:readSetting("footnote_popup_bottom_padding_size"))
+        padding_bottom = padding_bottom + padding_bottom_extra
+        htmlwidget_height = self.height - padding_top - padding_bottom + padding_bottom_extra
+    end
```
Comment (poire-z) [inline frontend/ui/widget/footnotewidget.lua:315]: You can probably just do to avoid the repetitions:
`local padding_bottom = Size.padding.large + Screen:scaleBySize(G_reader_settings:readSetting("footnote_popup_bottom_padding_size") or 0)`

**Comment by Commodore64user** [comment]
I was actually thinking about something like this but for slightly different reasons... Now as you can see in the following screenshot, I have left/right margins set to 0 (from the bottom menu), normally I have them at 10 but override the publisher margins however, in this case that was not possible; or perhaps it was possible but quite problematic as there are dozens of different paragraphs with different classes, and writing style tweaks for each was a pain...

anyway as you can see, since the pop-up seems to follow the margins we set, that monstrosity occurs

<p align="center">
 
</p> 

having some padding here on the left/right would also solve this. thank you for your attention to this matter.

**Comment by cookiebit** [comment]
> I would probably just get rid of the toggle, and just have, under "Footnote popup size": "Foonote bottom padding" which launches the SpinWidget with a default of 0 and an added button to reset to 0.
> It's not really something that once configured (if you need 60px on your phone) you will want to toggle.

That's a very good point! I originally started with only the toggle and a hardcoded padding value, but later changed that into making it adjustable, hence why the toggle exists in the first place. But your reasoning makes very much sense, so I'll look into removing the toggle altogether.

The approach with `Footnote popup size` -> `Footnote bottom padding` is also easier to adapt into having the option to set adjustable side padding as well, if that is a possibility. I'll have to look closer on the source to familiarize myself a bit more, this was first and foremost a proof-of-concept, since I can't say that I have been able to test edge cases and such, or whether this works as intended without side effects.

File: frontend/ui/widget/footnotewidget.lua (hunk 1 above)
Comment (cookiebit) [reply frontend/ui/widget/footnotewidget.lua:315]: Thanks! I haven't written very much Lua code since I wrote my thesis using LuaLaTeX, so I chose the "safe route", not what's probably considered best practice.

**Comment by poire-z** [comment]
> I have left/right margins set to 0 (from the bottom menu), normally I have them at 10 but override the publisher margins however, in this case that was not possible; or perhaps it was possible but quite problematic as there are dozens of different paragraphs with different classes, and writing style tweaks for each was a pain...

There is a "Ignore horizontal paragraphs margin/padding" styletweak.

> having some padding here on the left/right would also solve this. thank you for your attention to this matter.

Well, if it's only on one book every dozens, you can just live with it. No need to add an option that will be a pain to finely adjust each time to that book's publisher margins, as it would be a global setting (because if it's per-document, it will be even messier).

**Comment by Commodore64user** [comment]
> There is a "Ignore horizontal paragraphs margin/padding" styletweak.

yes I know, that is what I always use. In this case, there are way too many different p-tags with different classes which had different indentation and styles. There was no way of overcoming it

> you can just live with it

or we could just add this padding and problem solved. I don't mind having at zero and adjusting it when needed.

**Comment by cookiebit** [comment]
> I was actually thinking about something like this but for slightly different reasons... Now as you can see in the following screenshot, I have left/right margins set to 0 (from the bottom menu), normally I have them at 10 but override the publisher margins however, in this case that was not possible; or perhaps it was possible but quite problematic as there are dozens of different paragraphs with different classes, and writing style tweaks for each was a pain...
> 
> anyway as you can see, since the pop-up seems to follow the margins we set, that monstrosity occurs

> having some padding here on the left/right would also solve this. thank you for your attention to this matter.

It makes sense to me that it should be possible to set the left/right padding for the footnote widget independently from the reader margins, for exactly this reason. Adding left/right padding as well as _some_ bottom padding would probably be the best way to get the full text inside the rounded corners as well, without pushing the content too far from the bottom edge. Personally I'm a fan of having the option to customize as much as possible (within reason).

I did make changes to the code locally to see how it would work, and it seems to work as intended with some minor issues I need to investigate further:
 * When using pretty big values for the side margins, the scroll bar ends up being placed midway between the screen edge and the content. That could probably work, but personally i figured the best thing would be to have the scroll bar pushed away from the edge towards the text as well. That is, place the scroll bar in the left portion of the right margin.
 * With the approach I have taken so far, there ends up being a scroll bar of pretty much zero width when the margins are set to that value as well. This should be solvable, since the scrollbar width is scaled to fit the margin with narrow margins, and keep a given width (10px?) when there is plenty space.

I don't have the code handy at the moment, so I'll have to push it to the PR a little later.
